src/data/turbo-dynamodb-data-source.ts

```diff
@@ -0,0 +1,422 @@
+
+  const breaker = new CircuitBreaker<[DynamoTask<unknown>], unknown>(
+    async (...args: [DynamoTask<unknown>]) => {
+      const [task] = args;
+      return task();
+    },
+    {
+      timeout: process.env.NODE_ENV === 'local' ? 10_000 : 3000,
+      errorThresholdPercentage: 10,
+      resetTimeout: 30_000,
+    },
+  );
+
+  breaker.on('timeout', () =>
+    console.error('DynamoDB circuit breaker command timed out'),
+  );
+
+  const wrapper = {
+    fire<T>(task: DynamoTask<T>): Promise<T> {
+      return breaker.fire(task) as Promise<T>;
+    },
+    breaker,
+  };
+
+  dynamoBreakers.set(client, wrapper);
+  return wrapper;
+}
+
+function idToBinary(dataItemId: TransactionId): Uint8Array {
+  return Buffer.from(dataItemId, 'base64url');
+}
+
+/**
+ *  Diagram of offsets represented in DynamoDB offsets info
+ *              ┌────────────────────────────────────────────┐ <---- offset 0 in raw parent
+ *              │            Raw Parent Buffer               │
+ *              │  ┌──────────────────────────────────────┐  │
+ *              │  │           Parent Header              │  │
+ *              │  └──────────────────────────────────────┘  │
+ *              │  ┌──────────────────────────────────────┐  │ <---- payloadDataStart in raw parent
+ *              │  │          Parent Payload              │  │
+ *              │  │                                      │  │
+ *              │  │  ┌────────────────────────────────┐  │  │ <---- startOffsetInParentPayload in parent payload
+ *              │  │  │   Raw Nested Item Buffer       │  │  │       offset 0 in raw nested item
+ *              │  │  │                                │  │  │
+ *              │  │  │ ┌────────────────────────────┐ │  │  │
+ *              │  │  │ │    Nested Item Header      │ │  │  │
+ *              │  │  │ └────────────────────────────┘ │  │  │
+ *              │  │  │ ┌────────────────────────────┐ │  │  │ <---- payloadDataStart in raw nested item
+ *              │  │  │ │   Nested Item Payload      │ │  │  │       0 in nested item payload
+ *              │  │  │ │                            │ │  │  │
+ *              │  │  │ └────────────────────────────┘ │  │  │
+ *              │  │  └────────────────────────────────┘  │  │ <---- rawContentLength in raw nested item
+ *              │  └──────────────────────────────────────┘  │
+ *              └────────────────────────────────────────────┘
+ */
+
+export class TurboDynamoDbDataSource implements ContiguousDataSource {
+  private dynamoClient: DynamoDBClient;
+  private log: winston.Logger;
+  private circuitBreakerWrapper: {
+    fire<T>(task: DynamoTask<T>): Promise<T>;
+    breaker: CircuitBreaker<[DynamoTask<unknown>], unknown>;
+  };
+
+  constructor({
+    dynamoClient,
+    endpoint,
+    region,
+    credentials,
+    log,
+  }: {
+    dynamoClient?: DynamoDBClient;
+    endpoint?: string;
+    region?: string;
+    credentials?: {
+      accessKeyId: string;
```

Comment:
I assume we're planning on just `aws-vault` locally and instance role creds, so no real need to configure these, right?